### PR TITLE
Review `Optparser` library

### PR DESCRIPTION
**Pull request by MartyO256**
The `Optparser` library is used by Harpoon for specifying and parsing its CLI options.

This PR addresses a few non-functional issues encountered when reviewing the underlying code:

- Merging of the unchecked specification information for an option is somewhat inefficient as it requires a comparison for each of its fields. Instead, it is proposed that the steps for building such a specification be represented using a function rather than a record containing only the changes to be made. This reduces merging the information to sequentially applying the transforms as `let make transforms = transforms |> List.fold_left ( |> ) empty`.
- The `Optparser` library exports internal components. An `Optparser` main module is proposed to re-export only the modules, types and functions required by the user for specifying and parsing CLI options. This PR also improves on the documentation for better usability with the added `optparser.mli` file.
- In the `OptInfo` module, field names are reused for the `'a unchecked` and `'a checked` types. This will raise errors in the build process as we upgrade `dune` and `ocaml`. Namespaces are introduced to avoid type-driven record field names disambiguation.

### CODE REVIEW

File: src/optparser/optSpec.ml [hunk 1]
```diff
@@ -1,71 +1,79 @@
...
+      ; actual_argument_count : int
+      }
+  end
+
+  type t =
+    [ `Missing_mandatory_option of Option.option_error
+    | `Invalid_arguments_length of Argument.invalid_arguments_length_error
+    | `Argument_reader_failure of Option.option_error
+    | `Not_an_option of Option.option_error
+    ]
```
Comment (Ailrun) [inline src/optparser/optSpec.ml:28]: May I ask the reason for this change? (Normal variant to polymorphic variant)

**Review by Ailrun — COMMENTED** [review]
Overall, it looks good to me. Here are some comments:

File: src/optparser/optSpec.ml (hunk 1 above)
Comment (MartyO256) [reply src/optparser/optSpec.ml:28]: That error type is being used with `('a, 'e) result`. Using polymorphic variants allows for composable error-handling: https://keleshev.com/composable-error-handling-in-ocaml

I realize that composable errors are not actually used in `Optparser`. To support that, we would need to make these changes to the parser types:
```diff
(* parser.mli *)
- val parse : 'a OptSpec.t -> string list -> ('a, OptSpec.Error.t) result
+ val parse : 'a OptSpec.t -> string list -> ('a, [> OptSpec.Error.t]) result

(* optSpec.ml *)
- type +'a t =
+ type (+'a, 'error) t 
  { opt_tbl :
      (string, int option * (help_printer -> string list -> unit)) Hashtbl.t
-  ; comp_value : string list -> ('a, Error.t) result
+  ; comp_value : string list -> ('a, [> Error.t] as 'error) result
  ; mandatory_help_entries : HelpEntry.t list
  ; optional_help_entries : HelpEntry.t list
  }
```

File: src/optparser/optInfo.ml [hunk 2]
```diff
@@ -1,175 +1,89 @@
...
-    ; help_msg = fold ~none:spec1.help_msg ~some spec0.help_msg
-    ; default_argument = fold ~none:spec1.default_argument ~some spec0.default_argument
-    ; optional = fold ~none:spec1.optional ~some spec0.optional
+(** A module for functions to give information for an option. For example,
+    functions that configure
+
+    - the name of an option
+    - the help message of an option
+    - the fallback value of an optional option
+    - etc. are included.
```
[the remark this answers is not in the document]
Comment (MartyO256) [reply src/optparser/optInfo.ml]: OCamlformat broke that, nice catch.

File: src/optparser/list.ml [hunk 3]
```diff
@@ -1,14 +1,29 @@
...
-    | _ when n <= 0 -> []
+  let rec go n = function
+    | _ when n <= 0 ->
+        []
     | [] ->
-       go n ls
+        go n ls
     | x :: xs ->
-       x :: go (n - 1) xs
+        x :: go (n - 1) xs
```
[the remark this answers is not in the document]
Comment (MartyO256) [reply src/optparser/list.ml:10]: I'm using OCamlformat with the configuration found here: https://github.com/MartyO256/Beluga/blob/signature/.ocamlformat

**Pull request merged**